src/init.cpp

```diff
@@ -793,6 +793,31 @@ void InitParameterInteraction(ArgsManager& args)
             LogInfo("parameter interaction: -onlynet excludes IPv4 and IPv6 -> setting -dnsseed=0\n");
         }
     }
+
+    // We drop UPnP support but kept the arg as hidden for now to display a friendlier error to user who have the
+    // option in their config, and upgrade the setting to -natpmp. TODO: remove for version 30.0.
+    if (args.IsArgSet("-upnp")) {
+        bool overridden = args.SoftSetBoolArg("-natpmp", true);
+        LogWarning("Option '-upnp' is set but UPnP support was dropped in version 29.0.%s",
+                overridden ? " Setting '-natpmp' instead." : "");
+    }
+
+    // If dynamic settings are enabled, also replace it in settings.json.
+    fs::path path;
+    if (args.GetSettingsPath(&path, /* temp= */ false)) {
+        common::SettingsValue value = args.GetPersistentSetting("upnp");
```

Comment:
Slightly surprisingly to me, `GetPersistentSetting()` gets values from `bitcoin.conf`, so if you set `upnp=1` in bitcoin.conf, the branch below is executed, even if there's no upnp setting in `settings.json`. Maybe this is intended, but in that case `if (value.isTrue)` below fails, so natpmp does not get enabled, despite the warning that gets logged.

Reply:
No, that's not intended. i'll look into a way to query the json only.

Reply:
Why would `value.isTrue` fail? I suspect what will happen (still have to test) is that you'll keep `upnp=1` in the config and _also_ have `natpmp: 1` in settings. If the former is ignored and the latter used, that should be fine.

Reply:
Aha, `!value.IsBool()` for `upnp=1`

Reply:
The real problem here is that GetPersistentSetting also looks into the configuration file. i'll add a flag to skip that and this should work as intended.

Reply:
Alternatively it could directly check the value and if necessary manipulate `settings.rw_settings`, inside the `LockSettings` lock. This also keeps it atomic (not that we care here). But it also gives the clearest code and doesn't need API changes.